File: src/main/java/com/micahelias/util/Color.java
package com.micahelias.util;

import org.joml.Vector4f;

public class Color {


  public static final Color LIGHT_GREEN = new Color(85, 239, 196);
  public static final Color PINK_GLAMOUR = new Color(255, 118, 117);
  public static final Color DRACULA = new Color(45, 52, 54);
  public static final Color ELECTRON_BLUE = new Color(9, 132, 227);
  public static final Color MINT_LEAF = new Color(0, 184, 148);
  public static final Color BRIGHT_YARROW = new Color(253, 203, 110);
  public static final Color WHITE = new Color(0, 0, 0);

  // Members

  public float red;
  public float green;
  public float blue;
  public float alpha;


  public Color(float red, float green, float blue) {
    this.red = red;
    this.green = green;
    this.blue = blue;
    this.alpha = 1.0f;
  }


  public Color(float red, float green, float blue, float alpha) {
    this.red = red;
    this.green = green;
    this.blue = blue;
    this.alpha = alpha;
  }

  public Color(int red, int green, int blue) {
    this.red = red / 255.0f;
    this.green = green / 255.0f;
    this.blue = blue / 255.0f;
    this.alpha = 1.0f;
  }


  public Color(int red, int green, int blue, float alpha) {
    this.red = red / 255.0f;
    this.green = green / 255.0f;
    this.blue = blue / 255.0f;
    this.alpha = alpha;
  }


  // Gets the color in vector form
  public Vector4f toVector() {
    return new Vector4f(red, green, blue, alpha);
  }

  public float getRed() {
    return red;
  }

  public float getGreen() {
    return green;
  }

  public float getBlue() {
    return blue;
  }

}

File: mango/util/color.go
package util

import (
	"math"

	glm "github.com/go-gl/mathgl/mgl32"
)

type Color struct {
	glm.Vec4

	red   float32
	green float32
	blue  float32
	alpha float32
}

var WHITE Color = NewColorRGBf(1.0, 1.0, 1.0)
var BLACK Color = NewColorRGBf(0.0, 0.0, 0.0)
var DRACULA Color = NewColorRGBi(45, 52, 54)
var MINT_GREEN Color = NewColorRGBi(0, 184, 148)
var PINK_GLAMOUR Color = NewColorRGBi(255, 118, 117)
var ELECTRON_BLUE Color = NewColorRGBi(9, 132, 227)

func NewColorRGBf(red, green, blue float32) Color {
	color := Color{}

	color.red = red
	color.green = green
	color.blue = blue
	color.Vec4 = glm.Vec4{red, green, blue, 1.0}

	return color
}

func NewColorRGBAf(red, green, blue, alpha float32) Color {
	color := Color{}

	color.red = red
	color.green = green
	color.blue = blue
	color.alpha = alpha
	color.Vec4 = glm.Vec4{red, green, blue, alpha}

	return color
}

func NewColorRGBi(red, green, blue int) Color {
	color := Color{}

	color.red = float32(red) / 255.0
	color.green = float32(green) / 255.0
	color.blue = float32(blue) / 255.0
	color.Vec4 = glm.Vec4{color.red, color.green, color.blue, 1.0}

	return color
}

func NewColorRGBAi(red, green, blue int, alpha float32) Color {
	color := Color{}

	color.red = float32(red) / 255.0
	color.green = float32(green) / 255.0
	color.blue = float32(blue) / 255.0
	color.alpha = alpha
	color.Vec4 = glm.Vec4{color.red, color.green, color.blue, alpha}

	return color
}


func DarkenColor(color Color, amount float32) Color {
  red := color.red * (1.0 - amount)
  rFinal := math.Max(0.0, math.Min(1.0, float64(red)))  

  green := color.green * (1.0 - amount)
  gFinal := math.Max(0.0, math.Min(1.0, float64(green)))  

  blue := color.blue * (1.0 - amount)
  bFinal := math.Max(0.0, math.Min(1.0, float64(blue)))  


  return NewColorRGBAf(float32(rFinal), float32(gFinal), float32(bFinal), color.alpha)


}
